Add ensemble statistics preprocessor and 'groupby' option for multimodel (#673)
ESMValGroup/ESMValCore: doc/recipe/preprocessor.rst

   for resolutions of ``< 0.5`` degrees the regridding becomes very slow and
   will use a lot of memory.

.. _ensemble statistics:

Ensemble statistics
===================
For certain use cases it may be desirable to compute ensemble statistics. For

> example to prevent models with many ensemble member getting excessive weight in

Reviewer: ```suggestion
example to prevent models with many ensemble members getting excessive weight in
```

the multi-model statistics functions.

Theoretically, ensemble statistics are a special case (grouped) multi-model

> statistics. However, they should typically be computed earlier in the workflow.

Author: done

   for resolutions of ``< 0.5`` degrees the regridding becomes very slow and
   will use a lot of memory.

.. _ensemble statistics:

Ensemble statistics
===================
For certain use cases it may be desirable to compute ensemble statistics. For
example to prevent models with many ensemble members getting excessive weight in
the multi-model statistics functions.

Theoretically, ensemble statistics are a special case (grouped) multi-model
statistics. This grouping is performed taking into account the dataset tags

> `project`, `dataset`, `experiment`, and (if present) `sub_experiment`.

Reviewer 2: ```suggestion
``project``, ``dataset``, ``experiment``, and (if present) ``sub_experiment``.
```

However, they should typically be computed earlier in the workflow.
Moreover, because multiple ensemble members of the same model are typically more
consistent/homogeneous than datasets from different models, the implementation
is more straigtforward and can benefit from lazy evaluation and more efficient
computation.

The preprocessor takes a list of statistics as input:

.. code-block:: yaml

    preprocessors:
      example_preprocessor:
        ensemble_statistics:
          statistics: [mean, median]

This preprocessor function exposes the iris analysis package, and works with all
(capitalized) statistics from the :mod:`iris.analysis` package
that can be executed without additional arguments (e.g. percentiles are not
supported because it requires additional keywords: percentile.).

> Note that `ensemble_statistics` will not return the single model and ensemble files,

Reviewer 2: ```suggestion
Note that ``ensemble_statistics`` will not return the single model and ensemble files,
```

> only the requested ensemble statistics results.

Reviewer 2: Can you add the comment about the special preprocessor that can be used to get the enemble statistics and the raw datasets here (l.981f). I think this should go to the `ensemble_statistics` section.

of the combined models (``span: full``). The preprocessor sets a common time
coordinate on all datasets. As the number of days in a year may vary between
calendars, (sub-)daily data with different calendars are not supported.
The preprocessor saves both the input single model files as well as the multi-model

> results. In case of not wanting to keep the single model files, set

Reviewer 2: ```suggestion
results. In case you do not want to keep the single model files, set the
```

> parameter `keep_input_datasets` to `false` (default value is `true`).

Reviewer 2: ```suggestion
parameter ``keep_input_datasets`` to ``false`` (default value is ``true``).
```

          statistics: [mean, median]
          exclude: [NCEP]

Multi-model statistics now also supports a ``groupby`` argument. You can group by

> any dataset key (``project``, ``experiment``, etc.) or a combination of keys. You can

Reviewer: ```suggestion
any dataset key (``project``, ``experiment``, etc.) or a combination of keys in a list. You can
```

          statistics: [mean, median]
          exclude: [NCEP]

Multi-model statistics now also supports a ``groupby`` argument. You can group by
any dataset key (``project``, ``experiment``, etc.) or a combination of keys in a list. You can
also add an arbitrary 'tag' to a dataset definition and then groupby that tag. When
using this preprocessor in conjunction with `ensemble statistics`_ preprocessor, you
can groupby ``ensemble_statistics`` as well. For example:

.. code-block:: yaml

    datasets:
      - {dataset: CanESM2, exp: historical, ensemble: "r(1:2)i1p1", tag: 'group1'}
      - {dataset: CCSM4, exp: historical, ensemble: "r(1:2)i1p1", tag: 'group2'}

    preprocessors:
      example_preprocessor:
        ensemble_statistics:
          statistics: [median, mean]
        multi_model_statistics:
          span: overlap
          statistics: [min, max]
          groupby: [ensemble_statistics, tag]

>           exclude: [NCEP]

Author: I removed the `tag` from this example, below I think there is another example that already shows how the grouping works.

.. code-block:: yaml

    preprocessors:

>       multi_model_save_input:

Reviewer 2: Can you move this section up so it directly follows "...set the parameter `keep_input_datasets` to `false` (default value is `true`).?

        multi_model_statistics:
          span: overlap
          statistics: [mean, median]
          exclude: [NCEP]

>             multi_model_save_input:

Reviewer 2: ```suggestion
```

        multi_model_statistics:
          span: overlap
          statistics: [mean, median]
          exclude: [NCEP]
          keep_input_datasets: false

Multi-model statistics now also supports a ``groupby`` argument. You can group by
any dataset key (``project``, ``experiment``, etc.) or a combination of keys in a list. You can
also add an arbitrary tag to a dataset definition and then group by that tag. When
using this preprocessor in conjunction with `ensemble statistics`_ preprocessor, you
can group by ``ensemble_statistics`` as well. For example:

.. code-block:: yaml

    datasets:
      - {dataset: CanESM2, exp: historical, ensemble: "r(1:2)i1p1"}
      - {dataset: CCSM4, exp: historical, ensemble: "r(1:2)i1p1"}

    preprocessors:
      example_preprocessor:
        ensemble_statistics:
          statistics: [median, mean]
        multi_model_statistics:
          span: overlap
          statistics: [min, max]
          groupby: [ensemble_statistics]
          exclude: [NCEP]

This will first compute ensemble mean and median, and then compute the multi-model
min and max separately for the ensemble means and medians. Note that this combination

> will not save the individual ensemble members, only the ensemble and multimodel statistics results.

Reviewer 2: Move this paragraph to `ensemble_statistics`

Author: When `ensemble_statistics` is described, `multi_model_statistics` has not been described yet so I thought it would fit better in here. But whatever it can be moved

In case of wanting to save both individual ensemble members as well as the statistic results,
the preprocessor chains could be defined as:

.. code-block:: yaml
  preprocessors:
    everything_else: &everything_else
      area_statistics: ...
      regrid_time: ...
    multimodel:
      <<: *everything_else
      ensemble_statistics:

> variables:

Reviewer 2: The indentation is wrong here

    tas_datasets:
      short_name: tas
      preprocessor: everything_else
      ...
    tas_multimodel:
      short_name: tas
      preprocessor: multimodel
      ...

When grouping by a tag not defined in all datasets, the datasets missing the tag will
be grouped together. In the example below, datasets `UKESM` and `ERA5` would belong to the same

> group, while the other datasets would belong to either `group1` or `group2`

Reviewer 2: ```suggestion
group, while the other datasets would belong to either ``group1`` or ``group2``
```
